crates/sui-core/src/authority.rs

```diff
@@ -1858,6 +1868,66 @@ impl AuthorityState {
         Ok(new_epoch_store)
     }
 
+    fn check_system_consistency(&self) {
+        if self
+            .expensive_safety_check_config
+            .enable_epoch_sui_conservation_check()
+        {
+            if let Err(err) = self.check_system_consistency_impl() {
+                if cfg!(debug_assertions) {
+                    panic!("{}", err);
+                } else {
+                    // We cannot panic in production yet because it is known that there are some
+                    // inconsistencies in testnet. We will enable this once we make it balanced again in testnet.
+                    warn!("System consistency check failed: {}", err);
+                }
+            }
+        }
+    }
+
+    fn check_system_consistency_impl(&self) -> SuiResult {
+        let mut total_storage_rebate = 0;
+        let mut total_sui = 0;
+        for o in self.database.iter_live_object_set() {
+            match o {
+                LiveObject::Normal(object) => {
+                    total_storage_rebate += object.storage_rebate;
+                    // get_total_sui includes storage rebate, however all storage rebate is
+                    // also stored in the storage fund, so we need to subtract it here.
+                    total_sui += object.get_total_sui(&self.database)? - object.storage_rebate;
+                }
+                LiveObject::Wrapped(_) => (),
+            }
+        }
+        let system_state = self
+            .get_sui_system_state_object_during_reconfig()
+            .expect("Reading sui system state object cannot fail")
+            .into_sui_system_state_summary();
+
+        let storage_fund_balance = system_state.storage_fund_total_object_storage_rebates;
+        fp_ensure!(
+            total_storage_rebate == storage_fund_balance,
+            SuiError::from(
+                format!(
+                    "Inconsistent state detected at epoch {}: total storage rebate: {}, storage fund balance: {}",
+                    system_state.epoch, total_storage_rebate, storage_fund_balance
+                ).as_str()
+            )
+        );
+        const SUI_SUPPLY: u64 = 10_000_000_000_000_000_000; // 10B SUI in Mist.
```

Comment:
There's a `TOTAL_SUPPLY_MIST` defined here: https://github.com/MystenLabs/sui/blob/d1811bf1429db112bb1b681877e836d8f7c40b70/crates/sui-types/src/gas_coin.rs#L31
Let's reuse that.